### PR TITLE
Do not count the number of parens inside str or comment

### LINKED ISSUE #1777: Wrong parentheses formatting inside `#[should_panic(expected = "")]`
For example:
```rust
#[test]
#[should_panic(expected = "(")]
fn foo() {}
```
After formatting:
```rust
#[test]
#[should_panic(expected = "(" ])]
fn foo() {}
```
Which gives compile error.

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #1777.

### CODE REVIEW

File: src/visitor.rs [hunk 1]
```diff
@@ -720,9 +720,18 @@ impl Rewrite for ast::NestedMetaItem {
...
 fn count_missing_closing_parens(s: &str) -> u32 {
-    let op_parens = s.chars().filter(|c| *c == '(').count();
-    let cl_parens = s.chars().filter(|c| *c == ')').count();
-    op_parens.checked_sub(cl_parens).unwrap_or(0) as u32
+    let mut op_parens: u32 = 0;
+    let mut cl_parens: u32 = 0;
+    let mut inside_str = false;
+    for c in s.chars() {
+        match c {
+            '"' => inside_str = !inside_str,
```
Comment (nrc) [inline src/visitor.rs]: Could they also be inside comments?

Comment (topecongiro) [inline src/visitor.rs]: Yes, I think so.

**Comment by topecongiro** [comment]
Rebased with updates. The function got much complicated than I would like it to be, although I suppose we could not avoid the complexity when dealing with comments.

File: src/visitor.rs [hunk 2]
```diff
@@ -720,9 +720,51 @@ impl Rewrite for ast::NestedMetaItem {
...
 
 fn count_missing_closing_parens(s: &str) -> u32 {
-    let op_parens = s.chars().filter(|c| *c == '(').count();
-    let cl_parens = s.chars().filter(|c| *c == ')').count();
-    op_parens.checked_sub(cl_parens).unwrap_or(0) as u32
+    let mut op_parens: u32 = 0;
+    let mut cl_parens: u32 = 0;
+    let mut inside_str = false;
+    let mut inside_bullet_comment = false;
+    let mut inside_single_line_comment = false;
```
Comment (nrc) [inline src/visitor.rs]: These variables would probably be better encoded as a `State` enum, I think

File: src/visitor.rs [hunk 3]
```diff
@@ -720,9 +720,51 @@ impl Rewrite for ast::NestedMetaItem {
...
+            }
+            '*' if inside_bullet_comment && next_char.map_or(false, |c| *c == '/') => {
+                inside_bullet_comment = false;
+            }
+            '\n' if inside_single_line_comment => inside_single_line_comment = false,
+            '"' => {
+                inside_str = !inside_bullet_comment && !inside_single_line_comment && !inside_str
+            }
+            '(' => {
+                op_parens += if inside_str || inside_bullet_comment || inside_single_line_comment {
```
Comment (nrc) [inline src/visitor.rs]: These two clauses might be better in the form `if ... { op_parens += 1 }`.

**Comment by topecongiro** [comment]
Rebased with the changes reflecting inline comments.

**Comment by nrc** [comment]
Thanks!

**Pull request merged**